### PR TITLE
Add serving pre-compressed content section

### LINKED ISSUE #105: Please add brotli support
`AddOutputFilterByType BROTLI ...`

https://github.com/kjdev/apache-mod-brotli/issues/8

Thanks.

### LINKED ISSUE #134: Serve pre-compressed files
How about introducing a new section under web performance to allow apache to serve pre-compressed `.br` and/or `.gz` files?

In order to save CPU time on server side by skipping realtime compression, you can compress any test files lile `.html`, `.css`, `.js` etc in advance. This way reaction time for files to be delivered might improve, too.

I managed to get it to work in one of my .htaccess using https://httpd.apache.org/docs/2.4/mod/mod_deflate.html#precompressed as guideline. Would be happy to open a PR and talk about details.

### PR DESCRIPTION
**Pull request by midzer**
Hi again,
after https://github.com/h5bp/server-configs-apache/issues/134 I finally got free time to open a PR.

This PR is not finished yet. In fact it is working for my server setup, but I think it needs more input about how to combine `br` and `gz` section properly.

Until now only `css|js|html|svg|ics|json` precompressed files are served. I guess we should support all non-binary files here.

Thanks for your help and have a nice day
midzer

---
Fix #134 and fix #105 and close #113 as per https://github.com/h5bp/server-configs-apache/issues/105#issuecomment-363739736

### CODE REVIEW

File: src/web_performance/serving_pre-compressed_content.conf [hunk 1]
```diff
@@ -0,0 +1,57 @@
+# ----------------------------------------------------------------------
+# | Serving pre-compressed content                                     |
+# ----------------------------------------------------------------------
+
+<IfModule mod_headers.c>
+
+    # Serve brotli compressed CSS, JS, HTML, SVG, ICS and JSON files if they exist 
+    # and the client accepts br.
+    RewriteCond "%{HTTP:Accept-encoding}" "br"
+    RewriteCond "%{REQUEST_FILENAME}\.br" -s
```
Comment (LeoColomb) [inline src/web_performance/serving_pre-compressed_content.conf]: I would use the `-f` flag ([see docs](https://httpd.apache.org/docs/current/mod/mod_rewrite.html#rewritecond))

File: src/web_performance/serving_pre-compressed_content.conf [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+      Header append Content-Encoding br
+
+      # Force proxies to cache gzipped & 
+      # non-brotlied css/js files separately.
+      Header append Vary Accept-Encoding
+    </FilesMatch>
+
+</IfModule>
+
+<IfModule mod_headers.c>
```
Comment (LeoColomb) [inline src/web_performance/serving_pre-compressed_content.conf]: Same comments for this part! 😸

**Comment by midzer** [comment]
Thanks for your feedback, @LeoColomb 
On a live website this reworked section is working fine.

Help appreciated on writing tests for this new section. I think supporting more non-binary file types will be the next step in this PR.

**Comment by LeoColomb** [comment]
@midzer Thanks a lot for your update!
Much better! 👍 💯 
Just wondering, is there any reason to not resolve [this comment/question](https://github.com/h5bp/server-configs-apache/pull/136#discussion_r168749986)?

**Comment by midzer** [comment]
Ooops, I overlooked this one.
New `FilesMatch` regex is working properly, as far as I can tell.

**Comment by LeoColomb** [comment]
@midzer I've put some tests to check if everything is working as expected.
Mostly yes, but [some headers seems to be erased/bypassed](https://travis-ci.org/h5bp/server-configs-apache/jobs/364079595#L1936). Maybe related to `RewriteRule "\.xx\.br$" "-" [T=xx,E=no-gzip:1]` rules?

**Comment by midzer** [comment]
@LeoColomb Thanks for your support. I think we have to add `.br` and `.gz` in global `%FilesMatchPattern%` variable in order to pass tests for

* content-security-policy
* x-frame-options
* x-ua-compatible
* x-xss-protection

HTTP response header. But I don't know whether this has any unwanted side effects.

I'm unsure about `content-type` HTTP response header though. Maybe it helps to simply add `charset=utf-8` in `[T=...` field.

**Comment by LeoColomb** [comment]
Both changes (`FilesMatchPattern` & `[T=...`) should be safe enough to give a try. 😄 

Could you make these editions? Thx! 👍 🙂

**Comment by midzer** [comment]
Looks like `[T=text/javascript;charset=UTF-8,...` is not supported, although it should work like `AddType` directive stated at http://httpd.apache.org/docs/current/rewrite/flags.html.

[Here](https://stackoverflow.com/a/6884071) is a possible workaround, but I'm unsure whether it's the way to proceed.

**Comment by midzer** [comment]
Yeah, been a pleasure to work on this together :)

**Pull request merged**